### PR TITLE
dictdump: support HDF5 link preservation

### LINKED ISSUE #3213: extend dicttoh5 to handle links
In analogy to what we did for attributes (#3005) I need a syntax to specify links for dicttoh5 and h5todict. For example when doing data processing we often want to do this:

```python
{
    "@NX_class": "NXroot",
    "1.1": {
        "@NX_class": "NXentry",
        "instrument": {
            "@NX_class": "NXinstrument",
            "fx8_det0": {"@NX_class": "NXdetector", "data": [0, 1, 2, 3]},
        },
        "my_calc": {
            "@NX_class": "NXprocess",
            "configuration": {
                "@NX_class": "NXcollection",
                ">y": "../instrument/fx8_det0/data",  # SYNTAX FOR LINK
            },
        },
    },
}
```

So attributes start with "@" and links start with ">"? The library would automatically handle relative links and soft vs. external. For example `"scans/test.h5::/1.1/instrument/fx8_det0/data"` would be a relative external link and `"/1.1/instrument/fx8_det0/data"` would be an absolute softlink.

Before spending time on the implementation I'd like to know whether this idea finds enough support and what the appropriate syntax could be. For example `">linkname"` or `"linkname>"`.

### PR DESCRIPTION
**Pull request by woutdenolf**
closes #3213 

HDF5 link support:

 * `dicttoh5`: values `h5py.SoftLink` and `h5py.ExternalLink` are supported
 * `h5todict`: `dereference_links` option. `True` by default, otherwise the dict will contain `h5py.SoftLink` and `h5py.ExternalLink` instances.

HDF5 attribute support:

 * `h5todict`: `include_attributes` option. `False` by default, otherwise the dict will contain attributes as tuple keys.

Nexus wrappers: support `">"` notation and refactor

* `h5_to_nexus_dict`:
  * convert tuple keys to `"@"` key notation
  * convert `h5py.SoftLink` and `h5py.ExternalLink` values to `">"` key notation (value is a URL as a string)
* `nexus_to_h5_dict`:
  * convert `"@"` key notation to tuple keys
  * convert `">"` key notation (value is a URL as a string) to  `h5py.SoftLink` and `h5py.ExternalLink` values
  * add up-link support for soft link with the `">"` key notation (resolved to absolute link)
* `nxtodict`: same as `h5_to_nexus_dict(h5todict(ddict))`
* `dicttonx`: same as `dicttoh5(nexus_to_h5_dict(ddict))`

### CODE REVIEW

**Comment by woutdenolf** [comment]
@t20100 I need to do up-link support to dicttonx differently (parents should be derived and not an argument). The rest is done.

**Comment by woutdenolf** [comment]
Ok I fixed the up-link support. Should be good now.

File: silx/io/dictdump.py [hunk 1]
```diff
@@ -412,28 +413,124 @@ def h5todict(h5file, path="/", exclude_names=None, asarray=True):
...
+            h5name = path + "/" + key
+            # Preserve HDF5 link when requested
+            if not dereference_links:
+                lnk = h5f.get(h5name, getlink=True)
+                if is_link(lnk):
+                    ddict[key] = lnk
+                    continue
+
+            try:
+                h5obj = h5f[h5name]
```
Comment (woutdenolf) [inline silx/io/dictdump.py:440]: @t20100 I added this but I should remove it again? (see #3162)

Comment (woutdenolf) [reply silx/io/dictdump.py]: I removed it

File: silx/io/dictdump.py [hunk 2]
```diff
@@ -291,78 +287,83 @@ def dicttoh5(treedict, h5file, h5path='/',
...
-    h5path="/",
-    mode="w",
-    overwrite_data=False,
-    create_dataset_args=None,
-):
-    """
-    Write a nested dictionary to a HDF5 file, using string keys as member names.
-    The NeXus convention is used to identify attributes with ``"@"`` character,
-    therefor the dataset_names should not contain ``"@"``.
+def nexus_to_h5_dict(treedict, parents=tuple()):
```
Comment (t20100) [inline silx/io/dictdump.py:293]: Shouldn't this be private?

File: silx/io/dictdump.py [hunk 3]
```diff
@@ -291,78 +287,83 @@ def dicttoh5(treedict, h5file, h5path='/',
...
+            copy[key] = nexus_to_h5_dict(value, parents=parents+(key,))
+        else:
+            copy[key] = value
+    return copy
 
-    Example::
 
-        import numpy
-        from silx.io.dictdump import dicttonx
+def h5_to_nexus_dict(treedict):
```
Comment (t20100) [inline silx/io/dictdump.py:339]: Same, shouldn't it be private?

**Review by t20100 — APPROVED** [review]
LGTM.
Minor comments about making 2 internal functions private.

File: silx/io/dictdump.py (hunk 2 above)
Comment (woutdenolf) [reply silx/io/dictdump.py:293]: The functions `h5_to_nexus_dict` (returns a dict without special objects) and `nexus_to_h5_dict` (returns a dict with special objects) could be useful if someone want to serialize/deserialize?

Comment (t20100) [reply silx/io/dictdump.py:293]: OK, if that can be useful let's keep it, from looking at the code, I though that was only for internal use.

**Pull request merged**